### ISSUE TITLE
Lower complexity of functions + multithreading for caching

### ISSUE THREAD

**Issue by ShadesOT** [We-the-People-civ4col-mod/Mod#193]
Low level functions need to be of complexity O(1),  mid level functions of complexity O(n) and high level functions of complexity(n^2) max. Or the user expericence suffers like it does now. Turn ends take forever. Opening a city screen takes 2 to 4 seconds. ...

So we need to build caches. A lot of them. Caching is complex and takes time. So again the UX will suffer. Unless ... we use multithreading for caching. - It runs parallel to the thoughts of the user and the speed of the mouse movement of the user. Almost all modern cpus are multicore layouts. And unless the user is faster than the multithreaded caching algorithm all the UX will be instant. And if not, maybe we have to wait 0.5 sec untile the multithreading handle arrives. And ... well, it can hardly be worse than now.

Background:
Curently I am walking through the whole call stack in `doYields()`. I am not finished even, but it looks like `calculateNetYields(...)` will set a new record of O(n^5 + 3*(n^4)+ ... ).- Multiply that with the cities on the map and you get O(n^6+x*n^5+y*n^4+...).

Ah yeah ... and we need some sort of notificator, event, observer pattern. One that enforces consistency. Or the whole thing will not work If some code does not notify the city cache to recalculate itself, ... looking at the city screen will show wrong values.

**Comment by ShadesOT** [comment]
Question: In what way is multithreading dependent on
- Boost
- the windows sdk
- the c++ toolkit 2003?

**Comment by ShadesOT** [comment]
Good news: Boost 1.32.0 offers classes for multithreading
https://www.boost.org/doc/libs/1_32_0/doc/html/threads.html

**Comment by Nightinggale** [comment]
If you really want multithreading, [threadpool](http://threadpool.sourceforge.net/) works if you downgrade to version 0.2.3. It's certainly not a modern tool, but it's a lot easier to use than boost threads directly.

@ErikSandbraaten once talked about another way of multithreading, which should be even better. It's worth repeating here.

**Comment by ShadesOT** [comment]
Threadpool works on top of boost. Ok. 
Question1: Is threadpool already included in our mod project?
Question2: Why is it necessary to use threadpool v0.2.3?
Question3: Setup is as easy as dumping the threadpool header files into our DLLSources folder and including them in the code files?

**Comment by Nightinggale** [comment]
`
Question1: Is threadpool already included in our mod project?
`
No, because the time I tested it I ended up saving 0.4 seconds while waiting more than a minute for the AI. It also had other issues, meaning I never committed that experiment.

`
Question2: Why is it necessary to use threadpool v0.2.3?
`
Boost 1.37 changed the thread API and we have to use a threadpool written for the old API.

`
Question3: Setup is as easy as dumping the threadpool header files into our DLLSources folder and including them in the code files?
`
I think so. Also I think I ended up adding the library files as project source files because I had DLL linking issues. However be aware that when you create a new threadpool instance, it has overhead from creating and stopping threads. I wonder if we can make a global one and just reuse it to avoid this overhead.

**Comment by ErikSandbraaten** [comment]
I figures it was about time that people brought up multithreading.

Back in the summer\autumn of 2017 I began experimenting with using Intel TBB,  (https://www.threadingbuildingblocks.org/) which is a very popular fork-join template library, with the purpose of parallelizing various algorithms\game systems. TBB is preferable to naked threads (or pools) in the sense that it provides features like load-balancing, prevention of oversubscription\undersubscription etc. TBB is based on the concept of tasks, which are logical units of work rather than threads. This approach has many advantages. For example, tasks are potentially "infinitely scaleable" (given that you have enough resources) and they are composable, as opposed to theads. In addition, TBB offers pre-built algorithm patterns like parallel_for, parellel_reduce, parallel_template that greatly simplies the implementation of more complex concurrency patterns.

**Comment by ErikSandbraaten** [comment]
My efforts were based on RaR 2.6, so it needs to be forward-ported to 2.7.x
I need to check my notes, but I think I managed to parallelize the following:
- Job assignment
- Plot processing (primarily the time consuming visibility calculation, including the call chain to calculateNetYields and friends)
- Industrial victory condition (lot's of silly loops over yields)
- Achievements

**Comment by ErikSandbraaten** [comment]
The most common patterns is parallel_reduce, which is difficult to implement with a threadpool which is roughly equivalent to a non-scaleable parallel_for without load balancing.
Another benefit of TBB is that by using the patterns, the code looks more like using standard STL rather than a mess of synchronization primitives and other low level stuff. It's even possible to "spin out tasks" that run asynchronously with respect to the caller without having to deal with the hazzle of synchronizing stuff by dealing with condition variables. Arbitrarily complex dependency graphs can be created this way.

**Comment by ErikSandbraaten** [comment]
I'm sorry about writing this wall of text but I just wanted to warn you guys that threadpools \ naked threads are basically like programming assembly while Tasks are a high level construct that fits in naturally with C++ abstractions.

**Comment by ShadesOT** [comment]
Ok I read in the tbb docs.
**Likes**
- The [containers](https://software.intel.com/en-us/node/506076) can turn out to be very usefull.

**Dislikes**
- Task parallelization with parallel_for, parallel_reduce and parallel_do. Why? Because we would have to weave a lot of tbb-related code into the existing code. Eg every for loop would be affected whereas with normal threading we would change this example ...
```cpp
class City{
public:  
  void doTurn(){}
}

vector<City> vCities;

void doTurn()
{
  for(int i=0; i<vCities.length;i++)
  {
    doTurnCity(i);
  }
}
```
... simply like this ...
```cpp
class City{
public:  
  void doTurn(){}
}

vector<City> vCities;

void doTurn()
{
  for(int i=0; i<vCities.length;i++)
  {
    Threadpool.startThread(vCities[i], doTurn);
  }
  Threadpool.waitForAll();
}
```

TBB would require us to rewrite the code to something like this:
```cpp
class City{
public:  
  void doTurn(){}
}

vector<City> vCities;

class ApplyDoTurnCity {
    vector<City> vCityInternal;
public:
    void operator()( const TBB::blocked_range<size_t>& r ) const {
        vector<City> vCitiesLocal = vCitiesInternal;
        for( size_t i=r.begin(); i!=r.end(); ++i ) 
           vCitiesLocal[i].doTurn();
    }
    ApplyDoTurnCity (vector<City>vCities ) :
        vCitiesInternal(vCities)
    {}
};

TBB::parallel_for(TBB::blocked_range<size_t>(0,n), ApplyDoTurnCity (vCities));
```
... which is a heck of a complicated rewrite.

- In general TBB, which may be powerful, seems to require the implementation not just of a lot of complex code but also of complex design patterns (eg [this](https://software.intel.com/en-us/node/684814)). IMO this maybe a great thing if the code was written based on TBB from the beginning.

**Comment by ShadesOT** [comment]
I have a genaral question regarding multithreading:
Does multithreading automatically mean that multiple cores are facilitated?

**Comment by Nightinggale** [comment]
I wonder if multithreading is the right task to look into right now. Ideally we should work towards fixing everything for 2.7.1. Also it turns out that multiplayer is fairly broken right now. Multithreading will certainly not make it easier to make network games stay in sync.

More speed is good, but shouldn't we try to focus on making our first stable release before looking into luxury features like that?

`
Does multithreading automatically mean that multiple cores are facilitated?
`
That's the whole idea. Our program have no control over CPU cores or which one is used for gaming. In fact if you have full load in one thread only, this thread will be moved around as it heats up the CPU and by cycling it through the cores, each core is only heated 25% of the time (on a 4 core CPU), which is one of the reasons why single threaded performance is less affected by thermal throttling than multi threaded. Another being the obvious that more cores are cooled through the same heatsink.
All this happens automatically and as a programmer for a regular program/game (like none-OS stuff), you don't have to consider this. It just happens and you have no influence.

**Comment by ShadesOT** [comment]
I was thinking about multithreading, because I want to build more caches/precalculate values. To do this, the raw data does not have to be thread safe, as the caching thread only reads it. If the caching algo finishes and the raw data was changed meanwhile, then the cached values are invalid already and it simply starts over.

The following example illustrates what I want. It assumes, that there is only one thread which can change the raw data and which accesses the cached values.

```cpp
class Cache
{
protected:
	int * rawData;
	int * cachedData;

	bool valid;
	ThreadLock getterLock;
	ThreadLock validityLock;

	//O(n^5)
	void recalculate()
	{
		//...

		this->validityCheck();
	}

	void checkValidity()
	{
		this->validityLock.waitForReleaseAndLock();
		if (this->valid)
		{
			this->getterLock.release();
		}
		else
		{
			this->valid = true;
			ThreadPool.getGoingThread(this, recalculate);
		}
		this->validityLock.Release();
	}

public:
	//O(1)
	void changeRawData(int change)
	{
		this->getterLock.lock();
		this->validityLock.waitForReleaseAndLock();

		//... data change

		this->valid = false;
		this->validityLock.Release();
		this->validityCheck();
	}

	//O(1)
	int* getCachedValues() const
	{
		this->getterLock.waitForRelease();
		return this->cachedData;
	}
};
```

**Comment by ShadesOT** [comment]
The answer to all your fears is this:

> It assumes, that there is only one thread which can change the raw data and which accesses the cached values.

**Comment by ShadesOT** [comment]
Also, it is the whole point to lock the getter until the cache is valid again.

**Comment by ShadesOT** [comment]
To summarize my answers:
The main game thread manipulates raw game data and later needs chached/calculated values.

Now this is done sequential. The main thread manipulates raw data, does something else and then wants to access calculated values (eg when opening the city screen) depending on that manipulated raw data. Then it has to do the calculation and can continue.

In my code example the **one single main thread** manipulates the raw data and by doing so dispatches a worker thread that starts to do the calculation. The main thread leaves and does something else. Later it wants to open the city screen. If the worker thread has finished calculating, the main thread can access the cahced data immediately. If not, it has to wait.

It can only be faster than now. And it is not a race. The main thread always and only can continue if the calculations are completed. As in a sequential implementation.

**Comment by Nightinggale** [comment]
`
Also, it is the whole point to lock the getter until the cache is valid again.
`
This part is my main issue and the whole point to why I started writing about this issue. There are two valid scenarios for the variable in question.

1. all threads can or will use the get function. None use the set function
2. one thread use the set function and only that function is allowed to use the get function

1 obviously doesn't need a lock because they are all just reading. Order doesn't matter. 2 doens't need a lock because there is only one thread accessing the variable in question. This means the lock adds nothing and will just waste time.

The only case where the lock could make sense is if one thread calls get and another calls set. However then we have this scenario. Say we have int iValue = 1 and threads A and B.

1. A sets iValue to 2
2. B is stalled in a lock
3. B get iValue 2

or

1. B get iValue 1
2. A sets iValue to 2

You can't control which scenario you get and in network games each computer might differ on the result. If B reads 1 on one computer 2 on another and the AI does something only if iValue >= 2, then the AI will no longer do the same on both computers and we have a desync. Not even locks will prevent this.

*****

What we can use multithreading to is this:

* A sets iValue to 2
* B sets IPower to 42
* C sets iMoves to 0

The order doesn't matter. No need for locks and the game will stay in sync regardless of when each thread is executed.

*****

If we really want to make the game inconsistent as in accepting race conditions like this, then we will have to rewrite the AI code to setting one computer to handle the AI and then use the "normal" network code to distribute the results to the other computers. This way race conditions will not cause desyncs, but not only is it a really major task to rewrite the way the AI takes turns, it also means we lose the way we have it right now that if we load a savegame and hit end turn, the AI will do 100% the same each time, which is useful for bug hunting and bugfix testing. As awesome as it sounds to have the AI work on just one computer and distribute the AIs to multiple computers, I'm against this idea because it's worth a whole lot more to be able to get persistent gameplay when bughunting.

**Comment by ShadesOT** [comment]
I see the cause of our argument now: I have in mind that there is only one worker thread, but I forgot to add a flag about, if a worker was already dispatched. Without that it would be nonsense. You spotted the inconsistency right away.

Here is the completed code example. Maybe there is some bug more, but I want to illustrate what I have in mind and I am heading for. Multithreading for caching. And yeah, important condition: One cache is built by one worker thread only. :-)

And before Erik joins tomorrow and asks why I am implementing something, that might be available somewhere already: It is just an illustration of what I want! :-D

```cpp
class Cache
{
protected:
	int * rawData;
	int * cachedData;

	bool valid;
	ThreadLock getterLock;
	ThreadLock workerLock;
	bool workerIsIdle = true;

	//O(n^1mrd)
	void recalculate()
	{
		//do really long and complex building of cache here
		//calculates cached values from raw data ...
		for (int i = 0; i < 999999; i++)
		{
			//some random work ...
			this->cachedData = this->rawData;

			//maybe even check periodically,
			//if the partially built cache is already invalidated
			
			//no use of workerLock needed here ...
			if (!this->valid)
			{
				break;
			}
		}

		//when finished building the cache ...
		this->workerLock.waitForReleaseAndLock();//blocking
		this->workerIsIdle = true;
		this->workerLock.release();

		this->validityCheck();
	}

	void validityCheck()
	{
		this->workerLock.waitForReleaseAndLock();//blocking
		if (this->valid)
		{
			this->getterLock.release();
		}
		else
		{
			if (this->workerIsIdle)
			{
				this->workerIsIdle = false;
				this->valid = true;
				ThreadPool.dispatchThread(this, recalculate);
			}
		}
		this->workerLock.Release();
	}

public:
	//O(1)
	void changeRawData(int change)
	{
		this->getterLock.lock();//non blocking
		this->workerLock.waitForReleaseAndLock();//blocking

		//... raw data change
		this->rawData[0] = change;

		this->valid = false;
		this->workerLock.Release();
		this->validityCheck();
	}

	//O(1)
	int* getCachedValues() const
	{
		this->getterLock.waitForRelease();//blocking
		return this->cachedData;
	}
};

class Main
{
	void main()
	{
		Cache c = new Cache();

		c.changeRawData();

		//... now do something that can be faster 
		//or slower than building the cache in c.

		c.changeRawData() // again

		//... do something else that can be faster 
		//or slower than building the cache in c.
		
		//... and finally
		c.getCachedValues();
	}
};
``´

**Comment by Nightinggale** [comment]
For the second post in a row you posted while I was typing.

*****

`
Later it wants to open the city screen. If the worker thread has completed the calculations, the main thread can access the cached data immediately. If not, it has to wait until the worker thread has finished.
`
That's a fatal flaw in thinking. We shouldn't try to speed up the GUI by changing the DLL. Part of the reason is that the main culprit in the slow GUI is python and how python calls the exe to set up widgets. For all we know you might try to speed up something, which takes 0.3% of the total time. If we are unhappy with the GUI speed, we need to examine the python code first.

We should optimize based on the doTurn profiling results for the AI. If we can cache yields and reduce the time the AI spend figuring out colonies, we also reduce the time it takes to read the same data from the GUI. In other words if you optimize the AI, the GUI often follows.

When you consider building a cache, you need to figure out when you need to update it and when it is needed. If you add a lock to reading, will the reading lock take longer than just calculating each time you need the data? (Don't laugh. It's a valid question you should consider). How big is the risk that you end up with an invalid cache without detecting it and hence introduce a bug? Before making major performance changes, it's a good idea to profile to be sure that the code you want to change is actually a bottleneck.

Let's assume you are right and it's a good idea to cache this (I have a hunch you are on to something regarding identifying slow code, but we need to test to be sure).

We basically have 3 options:

1. update cache in another thread
2. update cache in the same thread
3. mark cache dirty and recalculate the next time get is called

Which is the fastest? 1 adds an overhead to get, meaning it's slower than 2 and a bad idea if called too frequently. 2 is slow when updating, but is the fastest when calling. 3 is the fastest if the cache can be invalidated multiple times between it being read (if this happens with each get, then why cache?). It's also worth adding that if 1 needs the result relatively fast, updating the cache will be as slow as 2 because the thread stalls, in which case you gained nothing for the slowdown of checking lock in the getter.

We have an old (yet still good) example of a cache of type 2 in the code. It caches yield requirements for professions. It's called countless times each turn, but it only needs updating when starting/loading the game and when gaining a new Founding Father (and not even with each FF). Obviously this one has to be optimized for fast reading.

I fear you are trying to make type 1 when type 2 would be the correct one, though type 3 should also be considered. You might speed up the colony GUI while slowing down the AI.

*****

If this is about reading the yields produced on a plot or possible production from the plot, then I have optimization plans for that. In the CivEffect branch, the game starts by making a list of YieldTypes, which are possible to sell in the domestic market. When handling domestic sales, it loops the list and then handle the yields it reads from it. This is obviously faster as it skips every single yield we know will be 0.

Is the domestic market a major slowdown? Not particularly (and certainly not now), but it was a tiny piece of code with good GUI exposure, which made it the ideal for proof of concept and to test the classes involved in generating the list. Now that the "engine" for this setup works, we can move on to do the same for yields produced on plots. This means when filling the YieldArray we can skip any yields we can't produce on plots, such as cigars. Sure it doesn't use buzzwords like multithreading, but it greatly reduces the amount of code to execute and amount of data to look up from memory.

Is yields from plot a slow part of the code? Yes and it's a good place to optimize, both for the colony screen and the AI.

*****

I feel like I might come through as a grumpy old man saying "in my days we only had one CPU. Why use more?". However that's not the case. I would love to speed up the game by using more CPU cores.

The problem I have with multithreading is it's not free. There are countless of examples where turning single threaded into multithreaded caused a slowdown because the usage of locks makes performance worse than just running everything in a single thread and not care about locks. Another reason is it makes debugging harder and we have the issue of keeping network games in sync. We aren't even doing a great job at the last part even before adding more threads.

In other words I'm fine with adding more threads, but only if it adds noteworthy performance and it should not be at the price of stability. The reason I object is that I sense a high risk that this ticket is on the path of no performance gain (possibly even a slowdown) and to start out with it looked like it would be a minefield for network stability, though that part might be less of an issue now.

**Comment by ErikSandbraaten** [comment]
@ShadesOT 
I agee that TBB is more verbose than the threadpool option since we don't have lambdas (C++)
However, how to you express (map)reductions, those operations are the most common in AI code.

From my experience with multithreading the DLL, there are actually relatively few systems that will benefit at the moment. We even risk reducing performance due to the locking and synchroniation involved. The big sell-case for multithreading (regardless of approach) is speeding up the (UNIT) AI which would also enable more sophisticated evaluations and thus a "smarter" AI. However, this is currently blocked by us having to use the pathfinder that is implemented in the EXE. Thus I think that priority should be given to merging in the K-Mod pathfinder (since it fixes many bugs and also allows us to have several pathfinders running in parallel).

It's extremely important to spend time analyzing the profiling results and you will probably be surprised about what is time consuming and what is not. Even if something looks very  inefficient, it may be called so rarely that it's not worth bothering spending time optimizing it.

I also generally agree with @Nightinggale 's concerns.

**Comment by Nightinggale** [comment]
I started a discussion ticket on how to profile consistently #210.

Also I sense some of the "multi core code is faster" hype here. It can be and it can be a lot slower. Here is some interesting reading/viewing on the topic:
- [Make Your Program Slower With Threads](https://brooker.co.za/blog/2014/12/06/random.html)
- [Multiprocessor behaviour - what can't you rely on? - Peter Sewell](https://www.youtube.com/watch?v=FRbkvSDTfdA)

Both gives some insights into issues regarding going multitasking. It should also make more sense why I used the example where some function is called for each unit, meaning units are executed in parallel and I set the requirement that each unit can only write to CvUnit (itself) and read from everything else except not even read access to other CvUnit instances. This should remove races and the need for locks.

**Comment by Nightinggale** [comment]
We went with TBB for multi-threading rather than anything mentioned here. No need to keep this as an open task.
